source/common/http/header_utility.cc

```diff
@@ -243,13 +252,28 @@ bool HeaderUtility::isConnect(const RequestHeaderMap& headers) {
   return headers.Method() && headers.Method()->value() == Http::Headers::get().MethodValues.Connect;
 }
 
+bool HeaderUtility::isConnectUdp(const RequestHeaderMap& headers) {
+  return headers.Upgrade() &&
+         headers.Upgrade()->value() == Http::Headers::get().UpgradeValues.ConnectUdp;
+}
+
 bool HeaderUtility::isConnectResponse(const RequestHeaderMap* request_headers,
                                       const ResponseHeaderMap& response_headers) {
   return request_headers && isConnect(*request_headers) &&
          static_cast<Http::Code>(Http::Utility::getResponseStatus(response_headers)) ==
              Http::Code::OK;
 }
 
+bool HeaderUtility::rewriteAuthorityForConnectUdp(RequestHeaderMap& headers) {
```

Comment:
This function has two issues:
1. it's missing the required percent-decoding
2. it's not adding the `[` and `]` required when using IPv6 addresses in URLs (see definition of [`authority`](https://www.rfc-editor.org/rfc/rfc3986#section-3.2))

Both of these would cause IPv6 literals to fail, are those covered in tests?

Reply:
Thank you for catching them. I will send an additional commit soon. One problem of the existing test cases is that it doesn't really make use of the parsed host and port. Envoy always just connects to a configured test cluster. I reality, a Dynamic Forward Proxy Filter & Cluster should be configured for this. I will think about some test options and cover it in the test cases.

Reply:
I fixed the bug and added relevant unit test and integration test cases. Done.